Update documentation (#24)
nest/nest-extension-module: doc/synapse_models.rst

Writing synapse models
======================

.. info::

   For developing custom neuron and synapse models, please consider using the NESTML modeling language <https://nestml.readthedocs.org/>`_.

NEST has a very flexible system to allow users to write their own synapse types. Synapses in NEST can either have static parameters or apply some dynamics on them. Each connection needs to have at least the following parameters:

* The connection delay
* The connection weight
* The target node of the connection
* The receiver port, which identifies the connection on the postsynaptic side

> The source node of a connection is implicitly stored by the position in the data structure that is used internally. These parameters are implemented in the StaticConnection synapse type, which can be used as a base class for more advanced synapse types.  

Author: ```suggestion
These parameters are implemented in the static_synapse model, which can be used as a base class for more advanced synapse types.

Note that the source node of a connection is implicitly stored by the position in the data structure that is used internally in the NEST kernel.
```

Most NEST models are very simplistic in this respect so far: they differentiate at most between an excitatory and an inhibitory synapse, and all input with positive weight is handled by the excitatory, all with negative weight by the inhibitory synapse.

More channels can be handled using the receptor type/receiver port mechanism in NEST. Please see the ``iaf_psc_alpha_multisynapse`` NEST model for an example.

Temporal aspects
^^^^^^^^^^^^^^^^

Time is a very important issue when processing spikes. Especially when the effect of a spike on the neuron depends on the history or state of the neuron, it is crucial that one considers causality carefully.

This is the reason why spiking history information about a neuron is available only via the ArchivingNode interface. The paper [2]_ has more on spike history handling.

Spike handling
^^^^^^^^^^^^^^

NEST delivers spikes in batches. Simulation proceeds for min_delay (minimum delay) time. During this time, any generated spikes are stored in a central spike queue. After each min_delay interval, spikes are delivered from the central queue via the ConnectionManager and the Connection objects to the individual neurons.

> All delays are handled inside the neuron, as described above. This means that when a spike "passes through" the Connection object, the actual biological arrival time (time when spike occured at sender plus delay) of the spike may be up to max_delay (maximum delay) time units in the future. This means, in particular, that Connection objects cannot perform any computations that depend on the state of the neuron at the time of "biological" spike arrival; they can only use historic information.

Reviewer: ```suggestion
All delays are handled inside the neuron, as described above. This means that when a spike "passes through" the ``Connection`` object, the actual biological arrival time (time when spike occurred at sender plus delay) of the spike may be up to max_delay (maximum delay) time units in the future. This means, in particular, that ``Connection`` objects cannot perform any computations that depend on the state of the neuron at the time of "biological" spike arrival; they can only use historic information.
```

Author: merged with Jessica's suggestion

> All delays are handled inside the neuron, as described above. This means that when a spike "passes through" the Connection object, the actual biological arrival time (time when spike occured at sender plus delay) of the spike may be up to max_delay (maximum delay) time units in the future. This means, in particular, that Connection objects cannot perform any computations that depend on the state of the neuron at the time of "biological" spike arrival; they can only use historic information.

Author: ```suggestion
All delays are handled inside the neuron, as described above. This means that when a spike "passes through" the Connection object, the actual biological arrival time (time when spike occured at sender plus delay) of the spike may be up to ``max_delay`` (maximum delay) time units in the future. This means, in particular, that Connection objects cannot perform any computations that depend on the state of the neuron at the time of "biological" spike arrival; they can only use historic information.
```

> All delays are handled inside the neuron, as described above. This means that when a spike "passes through" the Connection object, the actual biological arrival time (time when spike occured at sender plus delay) of the spike may be up to max_delay (maximum delay) time units in the future. This means, in particular, that Connection objects cannot perform any computations that depend on the state of the neuron at the time of "biological" spike arrival; they can only use historic information.

Author: Merged with Pooja's suggestion
